src/sdc11073/entity_mdib/entity_consumermdib.py

```diff
@@ -0,0 +1,732 @@
+from sdc11073 import observableproperties as properties
+from sdc11073.exceptions import ApiUsageError
+from sdc11073.mdib.consumermdib import ConsumerMdibState
+from sdc11073.namespaces import QN_TYPE, default_ns_helper
+from sdc11073.xml_types import msg_qnames, pm_qnames
+
+from .entities import XmlEntity, XmlMultiStateEntity, get_xsi_type
+from .entity_consumermdibxtra import EntityConsumerMdibMethods
+from .entity_mdibbase import EntityMdibBase
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from lxml.etree import QName
+
+    from sdc11073.consumer.consumerimpl import SdcConsumer
+    from sdc11073.mdib.entityprotocol import EntityGetterProtocol, EntityTypeProtocol
+    from sdc11073.mdib.mdibbase import MdibVersionGroup
+    from sdc11073.pysoap.msgreader import ReceivedMessage
+    from sdc11073.xml_types.pm_types import CodedValue, Coding
+    from sdc11073.xml_utils import LxmlElement
+
+    from .entities import ConsumerEntityType, ConsumerInternalEntityType
+
+    XmlEntityFactory = Callable[[LxmlElement, str, str], ConsumerInternalEntityType]
+
+
+@dataclass
+class _BufferedData:
+    received_message_data: ReceivedMessage
+    handler: callable
+
+
+multi_state_q_names = (pm_qnames.PatientContextDescriptor,
+                       pm_qnames.LocationContextDescriptor,
+                       pm_qnames.WorkflowContextDescriptor,
+                       pm_qnames.OperatorContextDescriptor,
+                       pm_qnames.MeansContextDescriptor,
+                       pm_qnames.EnsembleContextDescriptor)
+
+
+def _mk_xml_entity(node: LxmlElement, parent_handle: str, source_mds: str) -> ConsumerInternalEntityType:
+    """Return a new XmlEntity or XmlMultiStateEntity.
+
+    This is the default consumer entity factory.
+    It creates one of ConsumerInternalEntityType, which are factories for ConsumerEntityType.
+    By using a different factory, the user can change this to use other classes.
+    """
+    xsi_type = get_xsi_type(node)
+    if xsi_type in multi_state_q_names:
+        return XmlMultiStateEntity(parent_handle, source_mds, xsi_type, node, [])
+    return XmlEntity(parent_handle, source_mds, xsi_type, node, None)
+
+
+class EntityGetter:
+    """Implementation of EntityGetterProtocol."""
+
+    def __init__(self, entities: dict[str, XmlEntity | XmlMultiStateEntity], mdib: EntityConsumerMdib):
+        self._entities = entities
+        self._mdib = mdib
+
+    def handle(self, handle: str) -> ConsumerEntityType | None:
+        """Return entity with given handle."""
+        try:
+            return self._mk_entity(handle)
+        except KeyError:
+            return None
+
+    def node_type(self, node_type: QName) -> list[EntityTypeProtocol]:
+        """Return all entities with given node type."""
+        ret = []
+        for handle, entity in self._entities.items():
+            if entity.node_type == node_type:
+                ret.append(self._mk_entity(handle))

```

Comment:
It turned out that this and the following getter methods need to use mdib_lock, because the iterable could change during iteration. Using yield would allow the user to block the lock for a long time, while returning a list would only hold the lock as short has possible.